Update the logcontext doc (#10353)
matrix-org/synapse: docs/log_contexts.md

so will `fun`.

It's all too easy to forget to `await`: for instance if we forgot that
`do_some_stuff` returned an awaitable, we might plough on regardless. This
leads to a mess; it will probably work itself out eventually, but not
before a load of stuff has been logged against the wrong context.
(Normally, other things will break, more obviously, if you forget to
`await`, so this tends not to be a major problem in practice.)

Of course sometimes you need to do something a bit fancier with your
awaitable - not all code follows the linear A-then-B-then-C pattern.
Notes on implementing more complex patterns are in later sections.

## Where you create a new awaitable, make it follow the rules

Most of the time, an awaitable comes from another synapse function.
Sometimes, though, we need to make up a new awaitable, or we get an awaitable
back from external code. We need to make it follow our rules.

The easy way to do it is by using `context.make_deferred_yieldable`. Suppose we want to implement
`sleep`, which returns a deferred which will run its callbacks after a
given number of seconds. That might look like:

```python
# not a logcontext-rules-compliant function

> 

Reviewer: ```suggestion
```

## Passing synapse deferreds into third-party functions

A typical example of this is where we want to collect together two or
more awaitables via `defer.gatherResults`:

```python
a1 = operation1()
a2 = operation2()

> a3 = defer.gatherResults(a1, a2)

Reviewer: ```suggestion
a3 = defer.gatherResults([a1, a2])
```

```

This is really a variation of the fire-and-forget problem above, in that
we are firing off `a1` and `a2` without awaiting on them. The difference
is that we now have third-party code attached to their callbacks. Anyway
either technique given in the [Fire-and-forget](#fire-and-forget)
section will work.

Of course, the new awaitable returned by `gather` needs to be
wrapped in order to make it follow the logcontext rules before we can
yield it, as described in [Where you create a new awaitable, make it
follow the
rules](#where-you-create-a-new-awaitable-make-it-follow-the-rules).

So, option one: reset the logcontext before starting the operations to
be gathered:

```python
async def do_request_handling():
    with PreserveLoggingContext():
        a1 = operation1()
        a2 = operation2()

>         result = await defer.gatherResults(a1, a2)

Author: ```suggestion
        result = await defer.gatherResults([a1, a2])
```

```

In this case particularly, though, option two, of using
`context.run_in_background` almost certainly makes more sense, so that
`operation1` and `operation2` are both logged against the original
logcontext. This looks like:

```python
async def do_request_handling():
    a1 = context.run_in_background(operation1)
    a2 = context.run_in_background(operation2)

>     with PreserveLoggingContext():

Reviewer: normally we'd use `make_deferred_yieldable` rather than `PreserveLoggingContext` here.

    listener_queue.clear()
```

So, both ends of the awaitable chain have now dropped their references,
and the awaitable chain is now orphaned, and will be garbage-collected at

> some point. Note that `await_something_interesting` is a generator
> function, and when Python garbage-collects generator functions, it gives
> them a chance to clean up by making the `yield` raise a `GeneratorExit`

Reviewer: ```suggestion
some point. Note that `await_something_interesting` is a coroutine, 
which Python implements as a generator function.  When Python
garbage-collects generator functions, it gives them a chance to 
clean up by making the `async` (or `yield`) raise a `GeneratorExit`
```

there is now nothing waiting for its return, so the request context is
never cleared.

To reiterate, this problem only arises when *both* ends of a awaitable
chain are dropped. Dropping the the reference to an awaitable you're

> supposed to be calling is probably bad practice, so this doesn't

Reviewer: ```suggestion
supposed to be awaiting is bad practice, so this doesn't
```

    return d
```

(in general, mixing `with LoggingContext` blocks and `async` in the same
function leads to slighly counter-intuitive code, under this scheme).

Because we leave the original `with` block as soon as the awaitable is
returned (as opposed to waiting for it to be resolved, as we do today),
the logcontext is cleared before control passes back to the reactor; so
if there is some code within `do_request_handling` which needs to wait
for an awaitable to complete, there is no need for it to worry about
clearing the logcontext before doing so:

```python
def handle_request():
    d = do_some_stuff()    # d is a Twisted Deferred
    d.addCallback(do_some_more_stuff)
    return d
```

--- and provided `do_some_stuff` follows the rules of returning an awaitable
which runs its callbacks in the original logcontext, all is happy.

The business of an awaitable which runs its callbacks in the original
logcontext isn't hard to achieve --- we have it today, in the shape of

> `context._PreservingContextDeferred`:

Reviewer: _PreservingContextDeferred was removed back in #2675.

I kinda think this whole section is just a source of confusion rather than providing any useful data now. I don't think the approach it's trying to outline carries over particularly well to `async`/`await`, so I don't think we'd take this approach today. WDYT about just removing it?

Author: > _PreservingContextDeferred was removed back in #2675.

Oh, good to know

> WDYT about just removing it?

Yeah agreed, let's do it

    return pcd
```

It turns out that, thanks to the way that awaitables chain together, we
automatically get the property of a context-preserving deferred with
`async`, provided the final Defered the function `yields` on has that
property. So we can just write:

```python
async def handle_request():
    await do_some_stuff()
    await do_some_more_stuff()
```

> To conclude: I think this scheme would have worked equally well, with

Reviewer: can you `s/yield/await` and `s/inlineCallbacks/async` in the "garbage-collection" section below? afaik it otherwise all still applies.

Author: Are you sure? It says it's because "`await_something_interesting` is a generator function" which it doesn't seem to be? Or is an awaitable just a fancy generator?

Reviewer: A coroutine is indeed, more-or-less, a fancy generator. Indeed, until Python 3.5, that's exactly what it was. The `async` syntax in Python 3.5 means that they aren't quite the same thing, but the mechanism is still basically the same.

After all, they amount to the same thing: a function gets halfway through, and is then suspended.
